### ISSUE TITLE
Add EXPERIMENTAL_changes functionality

### ISSUE THREAD

**Issue by mikedotexe** [near/near-api-js#390]
**Is your feature request related to a problem? Please describe.**
Many partners and members from the community want to use events. We currently send them here:
https://docs.near.org/docs/api/rpc-experimental#example-of-data-changes
While you can send such a response directly the RPC, `near-api-js` does not allow such a command to happen (as far as I'm aware) because we don't have the interface.

It seems you cannot use `.sendJsonRpc` until we map the parameters to the RPC request.
Until we do this, folks will have to use a JavaScript `fetch` or axios call.

**Describe the solution you'd like**
I believe we need to have an interface for `EXPERIMENTAL_changes` in:
`/src/providers/provider.ts`
that matches the RPC request struct (https://github.com/nearprotocol/nearcore/blob/master/core/primitives/src/rpc.rs)
```
pub struct RpcStateChangesRequest {
    #[serde(flatten)]
    pub block_id_or_finality: BlockIdOrFinality,
    #[serde(flatten)]
    pub state_changes_request: StateChangesRequestView,
}
}
```

**Describe alternatives you've considered**
Using `fetch` will work, but then they aren't using this library.

Lastly, we should revisit if this is truly "experimental" anymore.

**Comment by bowenwang1996** [comment]
`sendJsonRpc` is just a wrapper so you can pass the object you want to query with in there and it should work. Interface just makes it look nicer.

**Comment by mikedotexe** [comment]
> `sendJsonRpc` is just a wrapper so you can pass the object you want to query with in there and it should work. Interface just makes it look nicer.

I don't know how to do this

**Comment by vgrichina** [comment]
> > `sendJsonRpc` is just a wrapper so you can pass the object you want to query with in there and it should work. Interface just makes it look nicer.
> 
> I don't know how to do this

just pass plain JS object with necessary fields, taking into account serde annotations like https://serde.rs/attr-flatten.html

so in this case smth like:

```js

await provider.sendJsonRpc('EXPERIMENTAL_changes', { block_id: 1234567, changes_type: 'account_changes', account_ids: ['account1', 'account2'] });
```

**Comment by vgrichina** [comment]
@bowenwang1996 wonder if we can have generate TS types based on serde annotations? Then we won't really have to implement anything for most RPC calls.

**Comment by bowenwang1996** [comment]
That sounds feasible. @frol do you know any tool that does this?

**Comment by mikedotexe** [comment]
> > > `sendJsonRpc` is just a wrapper so you can pass the object you want to query with in there and it should work. Interface just makes it look nicer.
> > 
> > 
> > I don't know how to do this
> 
> just pass plain JS object with necessary fields, taking into account serde annotations like https://serde.rs/attr-flatten.html
> 
> so in this case smth like:
> 
> ```js
> await provider.sendJsonRpc('EXPERIMENTAL_changes', { block_id: 1234567, changes_type: 'account_changes', account_ids: ['account1', 'account2'] });
> ```

thanks, i'll give this a try. I had a hard time lining up which parameters matched as they seemed different on the RPC from the JS interface. Maybe I just confused myself. appreciate the code snippet

**Comment by frol** [comment]
@bowenwang1996 It is not allowed to use magic outside of Hogwarts. serde generates special serialization/deserialization code, and nothing more; there is no runtime introspection, and thus there should be a separate derive macro on every struct involved in the final response structure. Good news, https://github.com/wafflespeanut/paperclip can generate the code that will provide OpenAPI spec for the given struct and seems to be aware of serde attributes; bad new is that it cannot handle enums with inner values.

> Many partners and members from the community want to use events. We currently send them here: https://docs.near.org/docs/api/rpc-experimental#example-of-data-changes

@mikedotexe It is better to send them to https://github.com/nearprotocol/nearcore/tree/master/chain/indexer

**Comment by vgrichina** [comment]
@frol @bowenwang1996 I see, if there is no ready made tool – should be easier to code manually

**Comment by mehtaphysical** [comment]
Fixed in #534
